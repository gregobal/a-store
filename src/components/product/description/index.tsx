import {Typography} from "@alfalab/core-components/typography";
import {useMemo} from "react";

type Props = {
    description: string
}

const regex = /(?<=\.)\s+(?=[A-ZА-Я])/gm;

export const ProductDescription = ({description}: Props) => {
    const sentences = useMemo(() => {
        return description.split(regex);
    }, [description]);

    return (
        <>
            {sentences.map((sentence, idx) => (
                <Typography.Text
                    key={sentence}
                    tag="p"
                    defaultMargins={(idx + 1) % 3 === 0}
                    view="secondary-large"
                    weight="medium"
                >
                    {sentence}
                </Typography.Text>
            ))}
        </>
    );
}
