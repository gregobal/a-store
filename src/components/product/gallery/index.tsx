import {Button} from "@alfalab/core-components/button";
import {GenericWrapper} from "@alfalab/core-components/generic-wrapper";
import {Grid} from "@alfalab/core-components/grid";
import cn from "classnames";
import {useEffect, useState} from "react";
import {Image} from "../../image";

import styles from './index.module.css';

type Props = {
    images: string[]
    alt: string
}

export const ProductGallery = ({images, alt}: Props) => {
    const [current, setCurrent] = useState("");

    useEffect(() => {
        setCurrent(images[0]);
    }, [images]);

    const handleClick = (chosen: string) => {
        setCurrent(chosen);
    }

    return (
        <GenericWrapper column={true} grow={true}>
            <Image src={current} alt={alt} width={564} ratio="w3h4"/>
            <Grid.Row gutter={8} justify="left">
                {images.map((src) => (
                    <Grid.Col key={src} width="auto">
                        <Button
                            style={{backgroundImage: `url("${src}")`}}
                            className={cn(styles.button, {[styles.active]: src === current})}
                            view='link'
                            size="xl"
                            block={true}
                            onClick={() => handleClick(src)}
                        />
                    </Grid.Col>
                ))}
            </Grid.Row>
        </GenericWrapper>
    );
}
