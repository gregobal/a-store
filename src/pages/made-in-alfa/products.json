[
    {
      "id": 0,
      "preview": "http://qa-games.ru/astore/public/images/98058636.jpeg",
      "title": "Рюкзак «Для умных и свободных»",
      "price": 4999,
      "availability": true
    },
    {
      "id": 1,
      "preview": "http://qa-games.ru/astore/public/images/68519498.jpeg",
      "title": "Футболка Для умных и свободных",
      "price": 1999,
      "availability": true,
      "colors": ["white", "black", "#ef3124"]
    },
    {
      "id": 2,
      "preview": "http://qa-games.ru/astore/public/images/77117755.jpeg",
      "title": "Блокнот Для умных и свободных",
      "price": 1499,
      "availability": true
    },
    {
      "id": 3,
      "preview": "http://qa-games.ru/astore/public/images/15932051.jpeg",
      "title": "Чехол с кардхолдером",
      "price": 799,
      "availability": false
    },
    {
      "id": 4,
      "preview": "http://qa-games.ru/astore/public/images/56369345.jpeg",
      "title": "Экоручка",
      "price": 99,
      "availability": true
    }
  ]
