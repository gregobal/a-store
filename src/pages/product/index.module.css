@media (max-width: 1023px) {
	.container {
		align-self: center;
		max-width: 600px;
	}
}
