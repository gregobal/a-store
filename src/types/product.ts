export type Product = {
    id: number,
    preview: string,
    images: string[],
    title: string,
    subtitle?: string,
    description: string,
    price: number,
    availability: boolean,
    sizes?: string[],
    colors?: string[],
    models?: string[],
    stickerNumbers?: number[]
}
