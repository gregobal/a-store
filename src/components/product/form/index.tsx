import {Button} from "@alfalab/core-components/button";
import {Grid} from "@alfalab/core-components/grid";
import {BaseSelectChangePayload} from "@alfalab/core-components/select";
import {SelectResponsive} from "@alfalab/core-components/select/Component.responsive";
import {OptionShape} from "@alfalab/core-components/select/typings";
import {Space} from "@alfalab/core-components/space";
import {useMemo, useState} from "react";
import {Product} from "../../../types/product";

type Props = {
    product: Product
}

const optionKeys = ["colors", "sizes", "models", "stickerNumbers"] as const;
type OptionKey = typeof optionKeys[number];

type Options = Record<OptionKey, OptionShape[]>;

type Selected = Record<OptionKey, OptionShape>;

const labels: Record<OptionKey, string> = {
    colors: "Цвет",
    sizes: "Размер",
    models: "Модель",
    stickerNumbers: "Номер стикера"
}

export const ProductForm = ({product}: Props) => {
    const selectsOpts = useMemo(() => {
        return optionKeys.reduce((acc, key) => {
            const arr = product[key];
            if (arr && arr.length > 0) {
                return {
                    ...acc,
                    [key]: arr.map((v) => ({key: v, content: v}))
                };
            }
            return acc;
        }, {} as Options)
    }, [product]);

    const [selected, setSelected] = useState<Selected>(
        Object.entries(selectsOpts).reduce((acc, [key, value]) => ({
            ...acc,
            [key]: value[0]
        }), {} as Selected)
    );

    const handleChange = (key: OptionKey, {selected}: BaseSelectChangePayload) => {
        if (selected) {
            setSelected(prev => ({
                ...prev,
                [key]: selected
            }))
        }
    }

    const handleClick = () => {
    }

    return (
        <Grid.Row tag="section" gutter={0} justify="left">
            <Grid.Col tag="div" width={{mobile: 7, tablet: 5, desktop: 5}}>
                <Space size="m" fullWidth={true}>
                    {Object.entries(selectsOpts).map(([key, value]) => {
                        const optKey = key as OptionKey;
                        return (<SelectResponsive
                            key={key}
                            size="s"
                            block={true}
                            closeOnSelect={true}
                            label={labels[optKey]}
                            options={value}
                            selected={selected[optKey]}
                            onChange={(payload) => handleChange(optKey, payload)}
                        />);
                    })}
                    <Button view="primary" block={true} onClick={() => handleClick()}>
                        В корзину
                    </Button>
                </Space>
            </Grid.Col>
        </Grid.Row>
    );
}
