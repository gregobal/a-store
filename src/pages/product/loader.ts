import {LoaderFunctionArgs} from "@remix-run/router/utils";
import {getProduct} from "../../api";
import {Product} from "../../types/product";

export function productLoader({params: {id}}: LoaderFunctionArgs): Promise<Product> {
    if (!id) {
        throw new Error("Required param ':id' is undefined")
    }

    return getProduct(id);
}
