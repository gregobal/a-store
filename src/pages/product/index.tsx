import {Amount} from "@alfalab/core-components/amount";
import {Grid} from "@alfalab/core-components/grid";
import {Link} from "@alfalab/core-components/link";
import {Space} from "@alfalab/core-components/space";
import {Typography} from "@alfalab/core-components/typography";
import {useLoaderData, useNavigate} from "react-router-dom";
import {ProductDescription} from "../../components/product/description";
import {ProductForm} from "../../components/product/form";
import {ProductGallery} from "../../components/product/gallery";
import {useTitle} from "../../hooks/useTitle";
import {Product} from "../../types/product";

import styles from "./index.module.css";

const colWidth = {mobile: 12, tablet: 12, desktop: 6};

export const ProductPage = () => {
    const product = useLoaderData() as Product;
    const {title, price, images, description} = product;

    const navigate = useNavigate();

    useTitle(title);

    const handleNavigateBack = () => navigate(-1);

    return (
        <Grid.Row tag="article" gutter={24} justify="center" className={styles.container}>
            <Grid.Col tag="section" width={colWidth}>
                <ProductGallery images={images ?? []} alt={title}/>
            </Grid.Col>
            <Grid.Col tag="section" width={colWidth} className={styles.secondCol}>
                <Space size="m" fullWidth={true}>
                    <Typography.TitleResponsive tag="h1" view="small">
                        {title}
                    </Typography.TitleResponsive>
                    <Typography.TitleResponsive tag="div" view="xsmall" weight="bold">
                        <Amount
                            value={price}
                            currency='RUR'
                            minority={1}
                            bold="full"
                        />
                    </Typography.TitleResponsive>
                    <ProductForm key={product.id} product={product}/>
                    <ProductDescription description={description}/>
                    <Typography.Text view="primary-small">
                        <Link view="secondary" pseudo={true} onClick={handleNavigateBack}>
                            Вернуться назад
                        </Link>
                    </Typography.Text>
                </Space>
            </Grid.Col>
        </Grid.Row>
    );
}
