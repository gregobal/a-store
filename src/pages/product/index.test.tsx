import {render, screen, waitFor} from "@testing-library/react";
import {createMemoryRouter, RouterProvider} from "react-router-dom";
import {Product} from "../../types/product";

import products from "../made-in-alfa/_fixtures/products.json";
import {ProductPage} from "./index";

const product = products[0];
const loaderStub = () => new Promise<Product>((resolve) => resolve(product));

describe("Product Page", () => {
    it("snapshot", async () => {
        const view = render(
            <RouterProvider router={
                createMemoryRouter([{
                    path: "/:id",
                    element: <ProductPage/>,
                    loader: () => loaderStub()
                }], {
                    initialEntries: [`/0`]
                })}
            />
        );

        await waitFor(() => {
            expect(screen.getByText(product.title)).toBeInTheDocument();
        });

        expect(view).toMatchSnapshot();
    });
});
