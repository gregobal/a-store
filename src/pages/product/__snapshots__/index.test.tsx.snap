// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Page snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="grid__row_1gt5e grid__component_xpkmx grid__center_xpkmx grid__gutter-24_1gt5e container"
      >
        <section
          class="grid__col_1gt5e grid__component_17rb3 grid__width-mobile-12_17rb3 grid__width-tablet-12_17rb3 grid__width-desktop-6_17rb3"
        >
          <div
            class="generic-wrapper__component_1xe8s generic-wrapper__column_1xe8s generic-wrapper__grow_1xe8s"
          >
            <div
              class="skeleton__component_jp9h2 skeleton__animate_jp9h2 wrapper w3h4"
            >
              <img
                alt="Рюкзак «Для умных и свободных»"
                class="image"
                src=""
                width="564"
              />
            </div>
            <div
              class="grid__row_1gt5e grid__component_xpkmx grid__left_xpkmx grid__gutter-8_1gt5e"
            >
              <div
                class="grid__col_1gt5e grid__component_17rb3 grid__width-auto_17rb3"
              >
                <button
                  class="button__component_onnkc button__link_onnkc button__xl_onnkc button__component_1qjkj button__link_1qjkj button__block_onnkc button__iconOnly_onnkc button"
                  style="background-image: url(http://qa-games.ru/astore/public/images/15586005.jpeg);"
                  type="button"
                />
              </div>
              <div
                class="grid__col_1gt5e grid__component_17rb3 grid__width-auto_17rb3"
              >
                <button
                  class="button__component_onnkc button__link_onnkc button__xl_onnkc button__component_1qjkj button__link_1qjkj button__block_onnkc button__iconOnly_onnkc button"
                  style="background-image: url(http://qa-games.ru/astore/public/images/45862782.jpeg);"
                  type="button"
                />
              </div>
              <div
                class="grid__col_1gt5e grid__component_17rb3 grid__width-auto_17rb3"
              >
                <button
                  class="button__component_onnkc button__link_onnkc button__xl_onnkc button__component_1qjkj button__link_1qjkj button__block_onnkc button__iconOnly_onnkc button"
                  style="background-image: url(http://qa-games.ru/astore/public/images/75259815.jpeg);"
                  type="button"
                />
              </div>
            </div>
          </div>
        </section>
        <section
          class="grid__col_1gt5e grid__component_17rb3 grid__width-mobile-12_17rb3 grid__width-tablet-12_17rb3 grid__width-desktop-6_17rb3 secondCol"
        >
          <div
            class="space__spaceContainer_4n70i space__vertical_4n70i space__start_4n70i space__spaceContainerFullWidth_4n70i"
          >
            <div
              class="space__spaceItem_4n70i space__spaceItemFullWidth_4n70i"
              style="margin-bottom: 16px;"
            >
              <h1
                class="typography__component_h1nzy typography__styrene-small_qohs3 typography__medium_h1nzy"
              >
                Рюкзак «Для умных и свободных»
              </h1>
            </div>
            <div
              class="space__spaceItem_4n70i space__spaceItemFullWidth_4n70i"
              style="margin-bottom: 16px;"
            >
              <div
                class="typography__component_h1nzy typography__styrene-xsmall_qohs3 typography__bold_h1nzy"
              >
                <span
                  class="amount__component_96zmx amount__bold_96zmx"
                >
                  4 999
                  <span
                    class=""
                  >
                     ₽
                  </span>
                </span>
              </div>
            </div>
            <div
              class="space__spaceItem_4n70i space__spaceItemFullWidth_4n70i"
              style="margin-bottom: 16px;"
            >
              <section
                class="grid__row_1gt5e grid__component_xpkmx grid__left_xpkmx"
              >
                <div
                  class="grid__col_1gt5e grid__component_17rb3 grid__width-mobile-7_17rb3 grid__width-tablet-5_17rb3 grid__width-desktop-5_17rb3"
                >
                  <div
                    class="space__spaceContainer_4n70i space__vertical_4n70i space__start_4n70i space__spaceContainerFullWidth_4n70i"
                  >
                    <div
                      class="space__spaceItem_4n70i space__spaceItemFullWidth_4n70i"
                    >
                      <button
                        class="button__component_onnkc button__m_onnkc button__component_1qjkj button__primary_1qjkj button__block_onnkc"
                        type="button"
                      >
                        <span
                          class="button__text_onnkc button__stretchText_onnkc"
                        >
                          В корзину
                        </span>
                      </button>
                    </div>
                  </div>
                </div>
              </section>
            </div>
            <div
              class="space__spaceItem_4n70i space__spaceItemFullWidth_4n70i"
              style="margin-bottom: 16px;"
            >
              <p
                class="typography__paragraph_xcdcz typography__secondary-large_xcdcz typography__medium_xcdcz"
              >
                Поместится и ноутбук, и худи.
              </p>
              <p
                class="typography__paragraph_xcdcz typography__secondary-large_xcdcz typography__medium_xcdcz"
              >
                У рюкзака широкие красные лямки и светоотражающие элементы — вас заметят и днём, и ночью.
              </p>
            </div>
            <div
              class="space__spaceItem_4n70i space__spaceItemFullWidth_4n70i"
            >
              <span
                class="typography__primary-small_xcdcz"
              >
                <button
                  class="link__component_1r33z link__secondary_1b8mz link__pseudo_1r33z"
                  type="button"
                >
                  <span
                    class="link__text_1r33z"
                  >
                    Вернуться назад
                  </span>
                </button>
              </span>
            </div>
          </div>
        </section>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="grid__row_1gt5e grid__component_xpkmx grid__center_xpkmx grid__gutter-24_1gt5e container"
    >
      <section
        class="grid__col_1gt5e grid__component_17rb3 grid__width-mobile-12_17rb3 grid__width-tablet-12_17rb3 grid__width-desktop-6_17rb3"
      >
        <div
          class="generic-wrapper__component_1xe8s generic-wrapper__column_1xe8s generic-wrapper__grow_1xe8s"
        >
          <div
            class="skeleton__component_jp9h2 skeleton__animate_jp9h2 wrapper w3h4"
          >
            <img
              alt="Рюкзак «Для умных и свободных»"
              class="image"
              src=""
              width="564"
            />
          </div>
          <div
            class="grid__row_1gt5e grid__component_xpkmx grid__left_xpkmx grid__gutter-8_1gt5e"
          >
            <div
              class="grid__col_1gt5e grid__component_17rb3 grid__width-auto_17rb3"
            >
              <button
                class="button__component_onnkc button__link_onnkc button__xl_onnkc button__component_1qjkj button__link_1qjkj button__block_onnkc button__iconOnly_onnkc button"
                style="background-image: url(http://qa-games.ru/astore/public/images/15586005.jpeg);"
                type="button"
              />
            </div>
            <div
              class="grid__col_1gt5e grid__component_17rb3 grid__width-auto_17rb3"
            >
              <button
                class="button__component_onnkc button__link_onnkc button__xl_onnkc button__component_1qjkj button__link_1qjkj button__block_onnkc button__iconOnly_onnkc button"
                style="background-image: url(http://qa-games.ru/astore/public/images/45862782.jpeg);"
                type="button"
              />
            </div>
            <div
              class="grid__col_1gt5e grid__component_17rb3 grid__width-auto_17rb3"
            >
              <button
                class="button__component_onnkc button__link_onnkc button__xl_onnkc button__component_1qjkj button__link_1qjkj button__block_onnkc button__iconOnly_onnkc button"
                style="background-image: url(http://qa-games.ru/astore/public/images/75259815.jpeg);"
                type="button"
              />
            </div>
          </div>
        </div>
      </section>
      <section
        class="grid__col_1gt5e grid__component_17rb3 grid__width-mobile-12_17rb3 grid__width-tablet-12_17rb3 grid__width-desktop-6_17rb3 secondCol"
      >
        <div
          class="space__spaceContainer_4n70i space__vertical_4n70i space__start_4n70i space__spaceContainerFullWidth_4n70i"
        >
          <div
            class="space__spaceItem_4n70i space__spaceItemFullWidth_4n70i"
            style="margin-bottom: 16px;"
          >
            <h1
              class="typography__component_h1nzy typography__styrene-small_qohs3 typography__medium_h1nzy"
            >
              Рюкзак «Для умных и свободных»
            </h1>
          </div>
          <div
            class="space__spaceItem_4n70i space__spaceItemFullWidth_4n70i"
            style="margin-bottom: 16px;"
          >
            <div
              class="typography__component_h1nzy typography__styrene-xsmall_qohs3 typography__bold_h1nzy"
            >
              <span
                class="amount__component_96zmx amount__bold_96zmx"
              >
                4 999
                <span
                  class=""
                >
                   ₽
                </span>
              </span>
            </div>
          </div>
          <div
            class="space__spaceItem_4n70i space__spaceItemFullWidth_4n70i"
            style="margin-bottom: 16px;"
          >
            <section
              class="grid__row_1gt5e grid__component_xpkmx grid__left_xpkmx"
            >
              <div
                class="grid__col_1gt5e grid__component_17rb3 grid__width-mobile-7_17rb3 grid__width-tablet-5_17rb3 grid__width-desktop-5_17rb3"
              >
                <div
                  class="space__spaceContainer_4n70i space__vertical_4n70i space__start_4n70i space__spaceContainerFullWidth_4n70i"
                >
                  <div
                    class="space__spaceItem_4n70i space__spaceItemFullWidth_4n70i"
                  >
                    <button
                      class="button__component_onnkc button__m_onnkc button__component_1qjkj button__primary_1qjkj button__block_onnkc"
                      type="button"
                    >
                      <span
                        class="button__text_onnkc button__stretchText_onnkc"
                      >
                        В корзину
                      </span>
                    </button>
                  </div>
                </div>
              </div>
            </section>
          </div>
          <div
            class="space__spaceItem_4n70i space__spaceItemFullWidth_4n70i"
            style="margin-bottom: 16px;"
          >
            <p
              class="typography__paragraph_xcdcz typography__secondary-large_xcdcz typography__medium_xcdcz"
            >
              Поместится и ноутбук, и худи.
            </p>
            <p
              class="typography__paragraph_xcdcz typography__secondary-large_xcdcz typography__medium_xcdcz"
            >
              У рюкзака широкие красные лямки и светоотражающие элементы — вас заметят и днём, и ночью.
            </p>
          </div>
          <div
            class="space__spaceItem_4n70i space__spaceItemFullWidth_4n70i"
          >
            <span
              class="typography__primary-small_xcdcz"
            >
              <button
                class="link__component_1r33z link__secondary_1b8mz link__pseudo_1r33z"
                type="button"
              >
                <span
                  class="link__text_1r33z"
                >
                  Вернуться назад
                </span>
              </button>
            </span>
          </div>
        </div>
      </section>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
